### PR TITLE
Shared object for macos and linux

### PR DESCRIPTION
**Pull request by atbrakhi**
This PR produces shared objects for macOS and Linux. It was successfully tested on macOS. 
See https://github.com/servo/servo/issues/30593 

TODO:
* [x] shared object for angle. `.so`, `.dylib`
* [x] Feature flag

Co-authored-by: Mukilan Thiyagarajan [mukilan@igalia.com]

### CODE REVIEW

**Comment by atbrakhi** [comment]
I'm opening as draft to get some early feedback

**Comment by sagudev** [comment]
I think this should be behind cargo feature.

CI isn't happy about change, but that might not be PR issue.

**Comment by atbrakhi** [comment]
ok, so both macos and linux is passing on CI after [2bc03f1](https://github.com/servo/mozangle/pull/71/commits/2bc03f1238774eb79bfae4969d98a767a289adbf). There was an issue with linking libs that was causing test failure in both system. 
I was expecting to see some errors on windows! on it now

**Comment by atbrakhi** [comment]
Okay, so now we have everything behind a feature flag. This won't break the current behavior anymore.

I plan to send a separate PR for Windows, as it might take some time for me to get my hands on a Windows system. Meanwhile, we can review this PR.

cc @sagudev

File: build.rs [hunk 1]
```diff
@@ -195,39 +207,67 @@ fn build_angle(target: &String, egl: bool) {
...
             &["gfx/angle/checkout/src/common/system_utils_win.cpp"][..],
         ),
-    ] {
-        if target.contains(os) {
-            for source in sources {
-                build.file(source);
+    ];
+
+    let out_dir = env::var("OUT_DIR").unwrap();
+    let out_path = Path::new(&out_dir);
```
Comment (sagudev) [inline build.rs:213]: I would merge this and only keep out_path.

File: build.rs [hunk 2]
```diff
@@ -195,39 +207,67 @@ fn build_angle(target: &String, egl: bool) {
...
         }
-    }
 
-    for flag in &clang_args {
-        build.flag(flag);
-    }
+        println!("cargo:rustc-link-lib=dylib={}", "angle");
+        #[cfg(target_os = "macos")]
+        {
+            let file = out_path.join(format!("{}.dylib", "libangle"));
```
Comment (sagudev) [inline build.rs:239]: Redundant format!

**Review by sagudev — COMMENTED** [review]
If I understand this PR correctly, it only introduces dylib building but not reusing it right?

**Comment by atbrakhi** [comment]
@sagudev i think we are on same page.

so this is how i was planning to go with it:
- Create shared object for all system (windows is missing in this PR)
- afterwards reuse them in servo

**Comment by sagudev** [comment]
> @sagudev, would you like me to change anything in this PR?

It needs rebase because #69 landed, which changes how build.rs works, so I will review it again after rebase, but generally looks good.

File: build.rs [hunk 3]
```diff
@@ -198,6 +198,7 @@ fn build_lib(compiled_libraries: &mut HashSet<Libs>, target: &String, lib: Libs)
                 "darwin",
                 &[
                     "gfx/angle/checkout/src/common/system_utils_mac.cpp",
+                    "gfx/angle/checkout/src/common/system_utils_apple.cpp",
```
Comment (atbrakhi) [inline build.rs:201]: Looks like with new update of `mozangle`; `GetExecutablePath` was moved to `system_utils_apple.cpp`, so I had to add this file here.

Other than this, I have followed pretty much same steps we follow for building windows dll for `egl` and such. Also tested this version locally with servo. Works fine.

cc @sagudev

Comment (sagudev) [reply build.rs:201]: This looks right according to: https://github.com/servo/mozangle/blob/be871c1f5413a105d43390f7a7362b07e88abb50/gfx/angle/targets/angle_common/moz.build#L172

File: build.rs [hunk 4]
```diff
@@ -299,12 +299,41 @@ fn build_translator(compiled_libraries: &mut HashSet<Libs>, target: &String) {
             .flag_if_supported("-arch:SSE2"); // MSVC
     }
 
-    build.link_lib_modifier("-whole-archive");
```
Comment (sagudev) [inline build.rs:302]: Why is this removed?

I think this is/was needed for external uses of dlls. I can recheck this on my windows machine later today.

File: build.rs [hunk 5]
```diff
@@ -320,10 +349,6 @@ fn build_translator(compiled_libraries: &mut HashSet<Libs>, target: &String) {
         .clang_arg("c++")
         .clang_arg("-std=c++17");
 
-    if target.contains("x86_64") || target.contains("i686") {
-        builder = builder.clang_arg("-msse2")
-    }
-
```
Comment (sagudev) [inline build.rs:326]: ditto

File: build.rs (hunk 4 above)
Comment (atbrakhi) [reply build.rs:302]: hmm, I could be wrong. I didn't test on windows machine. Please let me know if it fails on windows machine, I will add it back. Same for the next comment!

Comment (sagudev) [reply build.rs:302]: Works for me in wr.

**Review by sagudev — APPROVED** [review]

**Comment by sagudev** [comment]
I do not have write access to this repo, so somebody else will need to merge this.

**Comment by jdm** [comment]
> I do not have write access to this repo, so somebody else will need to merge this.

Are you sure? You're part of the Developers team, which has write access.

**Comment by sagudev** [comment]
> > I do not have write access to this repo, so somebody else will need to merge this.
> 
> Are you sure? You're part of the Developers team, which has write access.

no access

**Pull request merged**